omroneipApp/src/drvOmroneip.cpp

```diff
@@ -519,1517 +590,1382 @@ asynStatus drvOmronEIP::optimiseTags()
+              driverName, functionName, asynIndex, drvUser->tagIndex, datatype.c_str());
   }
-  
-  // Get the datatype of the array
-  dtype = ss.substr(ss.find_last_of(' ')+1, ss.size()-(ss.find_last_of(' ')+1)-1); // We dont want the closing "
-  std::vector<std::string> singleExpandedData;
-  if (dtype =="UINT" || dtype =="INT" || dtype =="WORD" || dtype =="BOOL" || dtype =="UDINT" || 
-      dtype =="DINT" || dtype =="REAL" || dtype =="DWORD" || dtype =="ULINT" || dtype =="LINT" || 
-        dtype == "LREAL" || dtype =="LWORD" || dtype.substr(0,6) == "STRING")
+  else if (status == asynTimeout)
   {
-    singleExpandedData.push_back(dtype);
+    setParamStatus(asynIndex, asynTimeout);
+    asynPrint(pasynUserSelf, ASYN_TRACE_WARNING, "%s:%s Warn, Timeout occured while updating asyn parameter with asyn ID: %d tagIndex: %d Datatype %s\n",
+              driverName, functionName, asynIndex, drvUser->tagIndex, datatype.c_str());
   }
-  else
+  else if (status == asynSuccess)
   {
-    // We assume that we have an array of structs and so we call the expandStructsRecursive function to get the raw dtypes from the struct
-    std::vector<std::string> embeddedDtypes = expandStructsRecursive(rawMap, dtype);
-    singleExpandedData.push_back(dtype);
-    singleExpandedData.insert(std::end(singleExpandedData), std::begin(embeddedDtypes), std::end(embeddedDtypes));
-    singleExpandedData.push_back("end:struct"+dtype);
+    setParamStatus(asynIndex, asynSuccess);
+    setParamAlarmStatus(asynIndex, asynSuccess);
+    setParamAlarmSeverity(asynIndex, NO_ALARM);
   }
-
-  for (size_t i = 0; i<arrayLength; i++) {
-    expandedData.insert(std::end(expandedData), std::begin(singleExpandedData), std::end(singleExpandedData));
+  status = plc_tag_unlock(drvUser->tagIndex);
+  if (status != 0)
+  {
+    asynPrint(pasynUserSelf, ASYN_TRACE_ERROR, "%s:%s Err, Tag index: %d Error occured in libplctag while trying to unlock tag: %s\n",
+              driverName, functionName, drvUser->tagIndex, plc_tag_decode_error(status));
   }
-  return expandedData;
+  return;
 }
 
-std::vector<std::string> drvOmronEIP::expandStructsRecursive(structDtypeMap const& rawMap, std::string structName)
+void drvOmronEIP::readPoller()
 {
-  static const char *functionName = "expandStructsRecursive";
-  std::vector<std::string> row = rawMap.at(structName);
-  std::vector<std::string> expandedRow;
-  const std::string arrayIdentifier = "\"ARRAY[";
-  for (std::string dtype : row)
+  static const char *functionName = "readPoller";
+  std::string threadName = epicsThreadGetNameSelf();
+  std::string tag;
+  int status;
+  int timeTaken = 0;
+  double pollingDelay = 0; // To stop from overloading the PLC, we divide read requests throughout the polling interval
+  // The poller may not be fully initialised until the startPollers_ flag is set to 1, do not attempt to get the pPoller yet
+  while (!this->startPollers_ && !omronExiting_)
```

Comment:
Still not super-keen on while(sleep) loops. Would much prefer to see wait on events...